---
title: "Check JAX data"
author: "Bo Yu, Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)

library(viridis)
library(ggpointdensity)
library(dplyr)
library(data.table)
library(readxl)

theme_set(theme_classic())

personal_path = "/fh/working/sun_w/sshen/MorPhiC"
path = "/fh/fast/sun_w/MorPhiC/data/MorPhiC_exchange_experiment/"
dat_path = file.path(path, "DRACC-processed/JAX_exchange_experiment_DRACC_processed_August2025/Tables")

metadata_path = file.path(path, "JAX/morphic_exchange_experiment")

read_header <- function(file_name, sheet){
  
  meta_header = read_excel(file_name, sheet = sheet, 
                         skip = 3, n_max = 1, col_names = FALSE)
  meta_header = as.character(meta_header)
  meta_header = gsub("_cell_line", "", meta_header, fixed = TRUE)
  meta_header = gsub("differentiated_product.", "", meta_header, fixed = TRUE)
  meta_header = gsub(".text", "", meta_header, fixed = TRUE)
  
  meta_header
}
```

## Read in meta data

```{r fig.width = 3, fig.height=3}
meta_file = paste0(metadata_path, "/DPC_metadata_template_EGE_JAX-final.xlsx")

cl_header = read_header(meta_file, sheet = "Clonal cell line")
cl_header

cl = read_excel(meta_file, sheet = "Clonal cell line", skip = 5, col_names = FALSE)
names(cl) = cl_header
cl = as.data.frame(cl)
dim(cl)
cl[1:2,]

dc_header = read_header(meta_file, sheet = "Differentiated product")
dc_header

dc = read_excel(meta_file, sheet = "Differentiated product", 
                skip = 5, col_names = FALSE)
names(dc) = dc_header
dc = as.data.frame(dc)
dim(dc)
dc[1:2,]

table(dc$timepoint_value, useNA="ifany")

eas_header = read_header(meta_file, sheet = "Expression alteration")
eas_header

eas = read_excel(meta_file, sheet = "Expression alteration", 
                skip = 5, col_names = FALSE)
names(eas) = eas_header
eas = as.data.frame(eas)
dim(eas)
eas[1:2,]
```

## Check cell lines

```{r}
any(duplicated(cl$clonal.label))
table(cl$clonal.type)
table(cl$clonal.parental_name)
table(cl$clonal.zygosity)
sort(table(cl$clonal.clone_id), decreasing = TRUE)

table(dc$clonal.label %in% cl$clonal.label)
table(cl$clonal.label %in% dc$clonal.label)
setdiff(dc$clonal.label, cl$clonal.label)
setdiff(cl$clonal.label, dc$clonal.label)

any(duplicated(dc$label))
table(table(dc$clonal.label))
table(dc$timepoint_value)
table(dc$timepoint_unit)
table(dc$final_timepoint)
table(dc$differentiated_product_protocol_id, 
      dc$model_system)

table(dc$timepoint_value, 
      dc$model_system)
```

## Check Expression alteration strategy

```{r}
any(duplicated(eas$expression_alteration.label))
setdiff(eas$expression_alteration.label, cl$expression_alteration.label)
setdiff(cl$expression_alteration.label, eas$expression_alteration.label)

table(eas$expression_alteration.parent_protocol_id)
table(eas$expression_alteration.method)
table(eas$expression_alteration.genes.allele_specific)
table(eas$expression_alteration.parent_protocol_id, 
      eas$expression_alteration.genes.altered_gene_symbol, useNA="ifany")
table(eas$expression_alteration.parent_protocol_id, 
      eas$expression_alteration.genes.targeted_genomic_region, useNA="ifany")
table(eas$expression_alteration.genes.targeted_genomic_region, 
      eas$expression_alteration.genes.editing_strategy, useNA="ifany")
table(eas$expression_alteration.genes.editing_strategy, 
      eas$expression_alteration.genes.expected_rna_alteration_phenotype, useNA="ifany")
table(eas$expression_alteration.genes.editing_strategy, 
      eas$expression_alteration.genes.expected_protein_alteration_phenotype, useNA="ifany")
```

## Extract KO strategy information

```{r}
unique(dc$description)

dc <- dc %>%
  mutate(
    ko_strategy = case_when(
      str_detect(description, "knockout cell line, .* KO from JAX") ~ "KO_JAX",
      str_detect(description, "knockout cell line, .* KO from MSK") ~ "KO_MSK",
      str_detect(description, "^KOLF2\\.2J differentiated\\b") & str_detect(label, "^(JAX_|MSK_)") ~ "CTRL_KO_WT",
      str_detect(description, "^KOLF2\\.2J differentiated\\b") & str_detect(label, "^NW_") ~ "CTRL_AID_WT",
      str_detect(description, "^KOLF2\\.2J differentiated\\b") & str_detect(label, "^UCSF_") ~ "CTRL_CRISPRi_WT",
      str_detect(description, "auxin induced degron null") ~ "AID_NW",
      str_detect(description, "arboring OsTIR1 as a control for AID system") ~ "CTRL_AID_OsTIR1",
      str_detect(description, "derived DMSO control for degron null") ~ "CTRL_AID_DMSO",
      str_detect(description, "CRISPRi generated .* null") ~ "CRISPRi_UCSF",
      str_detect(description, "CRISPRi transgene in the AAVS1") ~ "CTRL_CRISPRi_AAVS1",
      str_detect(description, "derived DMSO control for CRISPR") ~ "CTRL_CRISPRi_DMSO",
      TRUE ~ NA_character_
    ),
    ko_gene = coalesce(
      str_match(description, "(?<=derived knockout cell line, )([A-Za-z0-9-]+)(?=\\s+KO\\b)")[,2],
      str_match(description, "(?<=degron null of )([A-Za-z0-9-]+)")[,2],
      str_match(description, "(?<=CRISPRi generated )([A-Za-z0-9-]+)(?=\\s+null\\b)")[,2]
    ),
    ko_gene = ifelse(is.na(ko_gene), "WT", ko_gene)
  )

table(dc$ko_strategy, dc$ko_gene, useNA = 'ifany')

table(substr(dc$description, 1, 53), dc$ko_strategy, useNA = 'ifany')
```


## Read in library preparation information

```{r}
lib_header = read_header(meta_file, sheet = "Library preparation")
lib_header

lib = read_excel(meta_file, sheet = "Library preparation", 
                skip = 5, col_names = FALSE)
dim(lib)
names(lib) = lib_header
lib = as.data.frame(lib)
dim(lib)
lib[1:2,]

any(duplicated(lib$label))
setdiff(dc$label, lib$label)
setdiff(lib$label, dc$label)
```

## Read in sequence file information

```{r}
seq_header = read_header(meta_file, sheet = "Sequence file")
seq_header

seq = read_excel(meta_file, sheet = "Sequence file", 
                skip = 5, col_names = FALSE)
names(seq) = seq_header
seq = as.data.frame(seq)
seq$file_id = gsub("_R(1|2)_001.fastq.gz", "", 
                   seq$sequence_file.label)
dim(seq)
seq[1:2,]

seq = unique(seq[,c("file_id", "library_preparation.label", 
                    "sequence_file.run_id")])
dim(seq)
seq[1:2,]
length(unique(seq$library_preparation.label))

table(seq$sequence_file.run_id)
```

## Merge tabs to make a flat metedata table

```{r}
intersect(names(dc), names(lib))
table(dc$label%in%lib$label, 
      useNA="ifany")
table(lib$label%in%dc$label, 
      useNA="ifany")

meta = merge(dc, lib)
dim(meta)
meta[1:2,]

intersect(names(meta), names(seq))

any(duplicated(meta$library_preparation.label))
any(duplicated(seq$library_preparation.label))

table(meta$library_preparation.label%in%seq$library_preparation.label, 
      useNA="ifany")
table(seq$library_preparation.label%in%meta$library_preparation.label, 
      useNA="ifany")

meta = merge(meta, seq)
dim(meta)
meta[1:2,]

update_name <- function(df1, old_name, new_name){
  ww1 = which(names(df1) == old_name)
  stopifnot(length(ww1) == 1)
  names(df1)[ww1] = new_name
  df1
}

meta = update_name(meta, "sequence_file.run_id", "run_id")
```

```{r}
# intersect(names(meta), names(cl))

# table(meta$clonal.label %in% cl$clonal.label)
# table(cl$clonal.label %in% meta$clonal.label)
# setdiff(meta$clonal.label, cl$clonal.label)
# setdiff(cl$clonal.label, meta$clonal.label)

# meta = merge(meta, cl)
# dim(meta)
# meta[1:2,]

table(meta$ko_strategy, useNA = "ifany")
meta$ko_strategy[which(is.na(meta$ko_strategy))] = "WT"

table(meta$ko_gene, useNA = "ifany")
meta$ko_gene[which(is.na(meta$ko_gene))] = "WT"

meta$model_organ = meta$model_system
meta$model_system = NA
meta$model_system = str_extract(meta$label, "^[^-]+")
table(meta$model_system, meta$model_organ)

table(meta$model_system, meta$ko_gene, useNA = "ifany")

fwrite(meta, file = file.path(personal_path, "JAX_meta_data.tsv"), sep="\t")
```

## Check count data

### genesCounts.csv

```{r fig.width = 5.5, fig.height=4}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts)-1):ncol(cts))]
any(is.na(cts))

setdiff(names(cts)[-1], meta$file_id)
setdiff(meta$file_id, names(cts)[-1])
```


```{r}
gc()
sessionInfo()
```
