---
title: "Differential expression analysis e"
author: "Bo Yu, Si Liu, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# explicitly set warning level in R, to prevent a warning from being converted to an error
options(warn=1)
```

Differential expression using the bulk RNAseq JAX_RNAseq08 data from the June 2025 release.

See bulk_RNAseq.xlsx for details of the dataset.

For PrS_day_6_nor_ASCL2_ELF5_NR2F2_nor, 20250327_GT25-RobsonP-14-run3, one of the two run_id values, has only one sample. The run_id covariate is ignored for the DE analysis.

## R libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(stringr)
library(MASS)
library(RColorBrewer)
library(DESeq2)
library(viridis)
library(ggpointdensity)
library(tidyverse)
library(data.table)
library(ComplexHeatmap)
library(UpSetR)
library(readxl)
library(ggsci)
library(ggprism)
theme_set(theme_classic())

path = "../../MorPhiC/data/MorPhiC_exchange_experiment"
dat_path = file.path(path, "DRACC-processed/JAX_exchange_experiment_DRACC_processed_August2025/Tables")
```

## Read in meta data

```{r fig.width = 3, fig.height=3}
meta = fread("data/JAX_meta_data.tsv", data.table = FALSE)
dim(meta)
meta[1:2, ]
names(meta)
names(meta)[apply(meta, 2, \(x) any(is.na(x)))]

table(meta$clonal.label)
table(table(meta$clonal.label))
table(table(meta$label))

meta$timepoint = as.factor(meta$timepoint_value)
```

## Read in gene count data

```{r}
cts = fread(file.path(dat_path, "genesCounts.csv"), data.table = FALSE)
dim(cts)
cts[1:2, c(1:2, (ncol(cts) - 1) : ncol(cts))]
# Align counts with meta data
stopifnot(
  setequal(meta$file_id, names(cts)[-1]),
  !any(duplicated(meta$file_id)),
  !any(duplicated(names(cts)[-1]))
)
meta = meta[match(names(cts)[-1], meta$file_id), ]
# To matrix
cts_dat = data.matrix(cts[, -1])
rownames(cts_dat) = cts$Name
```

### Exclude genes without annotation

```{r}
# Read in gene annotation
gene_anno = fread("../MorPhiC_bulk_RNAseq/data/gencode_v44_primary_assembly_info.tsv")
dim(gene_anno)
gene_anno[1:2, ]
# Make sure all KO genes have unique ensembl gene id
gene_anno |> 
  filter(hgnc_symbol %in% setdiff(meta$ko_gene, 'WT')) |>
  arrange(hgnc_symbol)
# Exclude genes without annotation
table(rownames(cts_dat) %in% gene_anno$ensembl_gene_id)
cts = cts[cts$Name %in% gene_anno$ensembl_gene_id, ]
cts_dat = cts_dat[rownames(cts_dat) %in% gene_anno$ensembl_gene_id, ]
```

### Discard genes whose expression is 0 in more than 75% of samples

```{r}
gene_info = data.frame(
  apply(cts_dat, 1, min), 
  apply(cts_dat, 1, median), 
  apply(cts_dat, 1, \(x) quantile(x, probs = 0.75)), 
  apply(cts_dat, 1, max)
)
gene_info[1:2, ]
names(gene_info) = c("min", "median", "q75", "max")
summary(gene_info)
table(gene_info$q75 > 0)

# Subset genes
w2kp = which(gene_info$q75 > 0)
cts_dat = cts_dat[w2kp, ]
gene_info = gene_info[w2kp, ]
```

### Check read depth vs. gene expression quantile

```{r, fig.width = 4, fig.height=3}
meta$read_depth = colSums(cts_dat)/1e6
meta$q75 = apply(cts_dat, 2, quantile, probs = 0.75)

my_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00", 
               "#FFFF33", "#A65628", "#F781BF", "#999999", "#66C2A5", "#FC8D62")
ggplot(meta, aes(x=read_depth, y=q75, color=ko_strategy, shape = ko_gene)) +
  geom_point(size=2, alpha=0.7) + ggtitle("Gene counts") + 
  scale_shape_manual(values = c(7, 16)) + 
  xlab("read depth (million)") + ylab("75 percentile of gene expression") + 
  labs(color = "KO type", shape = "KO gene") + 
  # scale_color_brewer(palette = "Set1")
  scale_color_manual(values = my_colors)
```

```{r fig.width = 5.2, fig.height=3}
unique(meta$run_id)
meta$run_id_short = meta$run_id
```


```{r}
# PCA of all samples
q75 = apply(cts_dat, 1, quantile, probs=0.75)
cts_dat_n = t(cts_dat[q75 > 0,])
cts_dat_n = log(median(meta$q75)*cts_dat_n/meta$q75 + 1)
dim(cts_dat_n)

sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
names(sample_pca)
pc_scores = sample_pca$x
eigen_vals = sample_pca$sdev^2
eigen_vals[1:5]/sum(eigen_vals)
pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
pvs

PC_df = data.frame(pc_scores)
PC_df = cbind(PC_df, meta)

gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=ko_gene, color=ko_strategy)) +
  geom_point(size=2.5, alpha=0.7) + 
  scale_color_manual(values = my_colors) +
  scale_shape_prism() + 
  labs(color="KO type", shape ="KO gene") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("Wild type samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)
```


## Check PCA result of WT samples

```{r fig.width = 6, fig.height=4.5}
sample2kp = which(meta$ko_gene == "WT")
cts_dat_m = cts_dat[,sample2kp]
meta_m    = meta[sample2kp,]
summary(meta_m$q75)
dim(cts_dat_m)

stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
table(meta_m$file_id == colnames(cts_dat_m), useNA="ifany")

q75 = apply(cts_dat_m, 1, quantile, probs=0.75)
cts_dat_n = t(cts_dat_m[q75 > 0,])
cts_dat_n = log(median(meta_m$q75)*cts_dat_n/meta_m$q75 + 1)
dim(cts_dat_n)

sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
names(sample_pca)
pc_scores = sample_pca$x
eigen_vals = sample_pca$sdev^2
eigen_vals[1:5]/sum(eigen_vals)
pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
pvs

PC_df = data.frame(pc_scores)
PC_df = cbind(PC_df, meta_m)

gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=ko_gene, color=ko_strategy)) +
  geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
  scale_shape_prism() + 
  labs(color="KO type", shape ="KO gene") + 
  xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
  ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
  ggtitle("Wild type samples") + 
  guides(
    color = guide_legend(title = NULL, order = 1),
    shape = guide_legend(title = NULL, order = 2)
  ) +
  theme(
    legend.margin = margin(0, 0, 0, 0), 
    legend.box.just = "left", 
    legend.direction = "vertical" 
  )

print(gPC)
```

## Analyze samples of different model systems or batches

```{r fig.width = 6, fig.height=4.5, message=FALSE, warning=FALSE}
meta$model_condition = paste(meta$model_system, meta$condition, sep="_")
table(meta$model_condition)

table(meta$run_id, meta$model_condition)

table(meta$run_id, meta$ko_gene)
table(meta$run_id, meta$gene_condition_group)
table(meta$ko_strategy, meta$ko_gene)

unique_model_condition_ss = unique(meta$model_condition)
unique_model_condition_ss = sort(unique_model_condition_ss)

for(cur_mc in unique_model_condition_ss){
  print("")
  print("==========================================")
  print(cur_mc)
  print("==========================================")
  print("")
  
  sample2kp = which(meta$model_condition == cur_mc)
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]
  
  # verify the consistency between ko_strategy from metadata column and 
  # that from filenames
  extracted_ko_strings = str_extract_all(meta_m$file_id, "PTC|WT|KOLF2")
  print(table(sapply(extracted_ko_strings, length)))
  print(table(meta_m$ko_strategy, unlist(extracted_ko_strings)))
  
  stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
  table(meta_m$file_id == colnames(cts_dat_m), useNA="ifany")
  

  q75 = apply(cts_dat_m, 1, quantile, probs=0.75)

  cts_dat_n = t(cts_dat_m[q75 > 0,])
  cts_dat_n = log(median(meta_m$q75)*cts_dat_n/meta_m$q75 + 1)
  dim(cts_dat_n)
  
  sample_pca = prcomp(cts_dat_n, center = TRUE, scale. = TRUE)
  names(sample_pca)
  pc_scores = sample_pca$x
  eigen_vals = sample_pca$sdev^2
  eigen_vals[1:5]/sum(eigen_vals)
  pvs = round(eigen_vals[1:5]/sum(eigen_vals)*100,1)
  pvs
  
  PC_df = data.frame(pc_scores)
  PC_df = cbind(PC_df, meta_m)
  
  colnames(PC_df)[which(colnames(PC_df)=="timepoint")] = "day"
  
  only_ko_strategy = setdiff(unique(PC_df$ko_strategy), "WT")

  unique_ko_genes = sort(unique(PC_df$ko_gene))
  unique_ko_genes = c(unique_ko_genes[unique_ko_genes!="WT"], "WT")
  PC_df$ko_gene = factor(PC_df$ko_gene, unique_ko_genes)
  
  print(table(PC_df$gene_condition_group, PC_df$ko_gene, useNA="ifany"))

  shapes = c(0, 1, 7, 8, 9, 10, 11, 15, 16, 17)
  
  
  table(meta_m$clonal.description, meta_m$timepoint)
  
  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=ko_gene, color=run_id_short)) +
              geom_point(size=2.5, alpha=0.7) + 
              scale_color_brewer(palette = "Dark2") +
              scale_shape_manual(values = shapes[1:length(unique(PC_df$ko_gene))]) + 
              xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
              ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
              ggtitle(paste0(cur_mc, " ", only_ko_strategy, " for knockout genes")) + 
              guides(
                color = guide_legend(title = NULL, order = 1),
                shape = guide_legend(title = NULL, order = 2)
              ) +
              theme(
                legend.margin = margin(0, 0, 0, 0), 
                legend.box.just = "left", 
                legend.direction = "vertical" 
              )

  print(gPC)
  
  gPC = ggplot(PC_df, aes(x = PC1, y = PC2, shape=ko_gene, color=gene_condition_group)) +
              geom_point(size=2.5, alpha=0.7) + scale_color_brewer(palette = "Dark2") +
                scale_shape_manual(values = shapes[1:length(unique(PC_df$ko_gene))]) + 
              xlab(sprintf("PC1 (%.1f%% variance)", pvs[1])) + 
              ylab(sprintf("PC2 (%.1f%% variance)", pvs[2])) + 
              ggtitle(paste0(cur_mc, " ", only_ko_strategy, " for knockout genes\n", 
                             "color shows gene-condition group the WT sample is for")) + 
              guides(
                color = guide_legend(title = NULL, order = 1),
                shape = guide_legend(title = NULL, order = 2)
              ) + 
              theme_classic() + 
              theme(
                legend.margin = margin(0, 0, 0, 0), 
                legend.box.just = "left", 
                legend.direction = "vertical" 
              )

  print(gPC)

}

```

Run DE analysis for each gene-condition level separately. 
They natually separate all batches, except for one sample from 20250327_GT25-RobsonP-14-run3, which presents no batch effect based on the PC plot. 

```{r}
table(meta$gene_condition_group, meta$run_id, meta$model_system, meta$timepoint)
```

For the output files, create two versions, one with direct information, one with padj set to NA for the genes when the number of 0 per test is >=4. 

In more details, for both versions of the outputs, make it one file per knockout gene per DE group (DE group can be model system, model system + condition, model system + run_id, model system + day, etc.). 

The first version of the output files contains 
    
    gene id, gene symbol

and for each knockout strategy:

    baseMean, log2FoldChange, lfcSE, stat, pvalue, padj
    
The second version of the output files contains  

    gene id, gene symbol, chr, strand, position

and for each knockout strategy:

    log2FoldChange, pvalue, padj (set to be NA if the number of 0 per test >= 4).

In addition, save out a separate file of the sample size for each comparison. This needs to be by DE group. 

### Prepare output directories for DE results

```{r message=FALSE, warning=FALSE}

release_name = "2025_06_JAX_RNAseq08"

output_dir = file.path("results", release_name)
raw_output_dir = file.path(output_dir, "raw")
processed_output_dir = file.path(output_dir, "processed")

if (!file.exists(raw_output_dir)){
  dir.create(raw_output_dir, recursive = TRUE)
}

if (!file.exists(processed_output_dir)){
  dir.create(processed_output_dir, recursive = TRUE)
}

```

### DE analysis for all genes
```{r fig.width = 4.8, fig.height=3.2, message=FALSE, warning=FALSE}

df_size = NULL

unique_model_ss = unique(meta$model_system)
unique_model_ss = sort(unique_model_ss)

for(model1 in unique_model_ss){
  
  print("")
  print("==========================================")
  print(model1)
  print("==========================================")
  print("")
  
  sample2kp = which(meta$model_system == model1)
  
  cts_dat_m = cts_dat[,sample2kp]
  meta_m    = meta[sample2kp,]
 
  table(meta_m$ko_strategy)
  stopifnot(all(meta_m$file_id == colnames(cts_dat_m)))
  
  ## Generate sample information matrix
  cols2kp = c("model_system", "condition", "ko_strategy", "ko_gene", "gene_condition_group", 
              "run_id", "run_id_short", "timepoint", "q75")
  sample_info = meta_m[,cols2kp]
  colnames(sample_info)[which(colnames(sample_info)=="timepoint")] = "day"
  
  dim(sample_info)
  sample_info[1:2,]
  
  sample_info$ko_gene_gene_condition_group = paste(sample_info$ko_gene, 
                                         sample_info$gene_condition_group, sep="_")
  
  print(table(sample_info$run_id, sample_info$ko_gene))
  print(table(sample_info$ko_gene, sample_info$day)) 
  print(table(sample_info$ko_gene_gene_condition_group, sample_info$day)) 
  
  # print the table of ko_gene and run_id by (day, condition) combinations
  sample_info$day_condition = paste0(sample_info$day, "_", sample_info$condition)
  unique_day_condition_model_s = unique(sample_info$day_condition)
  unique_day_condition_model_s = as.character(sort(unique_day_condition_model_s))
  
  for (cur_d_c in as.character(unique_day_condition_model_s)){
    sample_info_by_day = sample_info[which(sample_info$day_condition==as.character(cur_d_c)), ]
    print("---------------------------")
    print(paste0("Model system: ", model1))
    print(paste0("On day_condition ", as.character(cur_d_c), ": "))
    print("Cross table between knockout gene and run id:")
    print(table(sample_info_by_day$ko_gene, sample_info_by_day$run_id_short))    
    print("Cross table between knockout gene and knockout gene group:")
    print(table(sample_info_by_day$ko_gene, sample_info_by_day$gene_condition_group))  
  }

  sample_info$model_day = paste0(sample_info$model_system, 
                                 "_day_", 
                                 as.character(sample_info$day))
  
  sample_info$de_grp = sample_info$model_day
  
  if(model1 == "PrS"){ 
    sample_info$de_grp = paste0(sample_info$de_grp, "_", 
                                sample_info$condition, "_",
                                sample_info$gene_condition_group)
  }
  
  unique_de_grps = unique(sample_info$de_grp)
  sorted_de_grps = sort(unique_de_grps)
  
  for (d_group in sorted_de_grps){

    dg_index = which(sample_info$de_grp==d_group)
    cts_dat_m_dg = cts_dat_m[, dg_index]
    sample_info_dg = sample_info[dg_index, ]
    
    stopifnot(all(sample_info_dg$de_grp==d_group))
    
    print("")
    print("-----------------------------------")
    print(paste0("Model system: ", model1))  
    if (sum(sample_info_dg$ko_gene=="WT") < 2){
      print(sprintf("DE group %s has only %d wild type samples", 
                    d_group, sum(sample_info_dg$ko_gene=="WT")))
      print(sprintf("do not run DESeq2 for it"))
      print("-----------------------------------")  
      next
    }else{
      print(sprintf("DE group %s DESeq2 results:", d_group))
      print("-----------------------------------")  
    }    
    
    
    # do two steps of filtering
    # first, filter based on q75 of each gene
    # second, filter based on whether each gene is expressed in at least 4 cells
    
    dg_q75 = apply(cts_dat_m_dg, 1, quantile, probs=0.75)
    cts_dat_m_dg = cts_dat_m_dg[dg_q75 > 0,]
    
    print(sprintf("After filtering by gene expression q75, the number of genes reduces from %d to %d", 
                  length(dg_q75), nrow(cts_dat_m_dg)))

    n_pos = rowSums(cts_dat_m_dg>0)
    cts_dat_m_dg = cts_dat_m_dg[n_pos >= 4,]

    if (length(n_pos)==nrow(cts_dat_m_dg)){
      print("After filtering by nonzero expression in at least 4 samples, the number of genes does not change.")
    }else{
      print(sprintf("After filtering by nonzero expression in at least 4 samples, the number of genes reduces from %d to %d", 
                    length(n_pos), nrow(cts_dat_m_dg)))    
    }
    
    # update the q75 based on only genes that will be used in the DE analysis
    sample_info_dg$q75 = apply(cts_dat_m_dg, 2, quantile, probs = 0.75)
    sample_info_dg$log_q75 = log(sample_info_dg$q75)
    
    if (d_group == 'PrS_day_6_nor_ASCL2_ELF5_NR2F2_nor') {
      sample_info_dg$run_id = 'edited'
      print('Ignoring run_id covariate for PrS_day_6_nor_ASCL2_ELF5_NR2F2_nor')
    }
    
    if (length(unique(sample_info_dg$run_id))==1){
      dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene + log_q75)
    }else{
      dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                    ~ ko_gene + run_id + log_q75)     
    }
    
    start.time <- Sys.time()
    dds = DESeq(dds)
    end.time <- Sys.time()

    time.taken <- end.time - start.time
    print(time.taken)
    
    ## association with read-depth
    res_rd = results(dds, name = "log_q75")
    res_rd = as.data.frame(res_rd)
    dim(res_rd)
    res_rd[1:2,]
    
    table(is.na(res_rd$padj))
  
    g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
      geom_histogram(color="darkblue", fill="lightblue", 
                     breaks = seq(0,1,by=0.01)) + 
      ggtitle(paste0(d_group, "\nread depth"))
    print(g0)
  
    ## Rerun the analysis without read-depth if it is not significant for 
    ## most genes, and the number of samples is small
    ## i.e., proportion of non-null in the distribution is smaller than 0.01
    ## (this needs to be restricted to the genes with not NA adjusted pvalue)
    ## or if smaller than 0.1 and the number of samples is not greater than 6

    pi_1_rd = max(0, mean(res_rd$pvalue[!is.na(res_rd$padj)] < 0.05) - 0.05)
    pi_1_rd = max(pi_1_rd, 0, 1 - 2*mean(res_rd$pvalue[!is.na(res_rd$padj)] > 0.5))
    
    print(sprintf("prop of non-null for rd: %.2e", pi_1_rd))
    
    if(pi_1_rd < 0.01 || (ncol(cts_dat_m_dg) <= 6 && pi_1_rd < 0.1 )){
      
      print("rerun DESeq2 without read depth")
      
      include_rd = 0
      
      if (length(unique(sample_info_dg$run_id))==1){
        dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                      ~ ko_gene)
      }else{
        dds = DESeqDataSetFromMatrix(cts_dat_m_dg, colData=sample_info_dg, 
                                      ~ ko_gene + run_id)     
      }
      
      dds = DESeq(dds)
      
    }else{
      include_rd = 1
    }

    ## association with run_id
      
    if (length(unique(sample_info_dg$run_id))>1){
      if(include_rd)
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene + log_q75)
      else
        dds_lrt = DESeq(dds, test="LRT", reduced = ~ ko_gene)
      
      res_lrt = results(dds_lrt)
    
      res_run_id = as.data.frame(res_lrt)
      dim(res_run_id)
      res_run_id[1:2,]
    
      table(is.na(res_run_id$padj))
      
      g0 = ggplot(res_run_id %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(d_group, "\nRun ID"))
      print(g0)
    
    }
    
    ## DE analysis for each KO gene
    table(sample_info_dg$ko_gene)
    
    genos = setdiff(unique(sample_info_dg$ko_gene), "WT")
    genos = sort(genos)
    
    for(geno in genos){
      
      res_g1 = results(dds, contrast = c("ko_gene", geno, "WT"))
      res_g1 = signif(data.frame(res_g1), 3)

      # add a column to record the number of zeros per test
      test_index = which(sample_info_dg$ko_gene%in%c(geno, "WT"))
      
      cts_dat_m_dg_test = cts_dat_m_dg[, test_index]
      n_zero = rowSums(cts_dat_m_dg_test==0)
      res_g1$n_zeros = n_zero    

      # record the number of samples involved in the comparison
      test_ko_gene_vec = sample_info_dg$ko_gene[test_index]
      
      df_size = rbind(df_size, 
                      c(d_group, 
                        paste0(geno, "_PTC"), 
                        sum(test_ko_gene_vec!="WT"), 
                        sum(test_ko_gene_vec=="WT")))   
      
      # prepare a processed version of output
      res_g1_reduced = res_g1[, c("log2FoldChange", "pvalue", "padj", "n_zeros")]
      res_g1_reduced$padj[which(res_g1_reduced$n_zeros>=4)] = NA
      
      if ((sum(test_ko_gene_vec!="WT")==1)|(sum(test_ko_gene_vec=="WT")==1)){
        res_g1_reduced$padj = NA
      }
      
      res_g1_reduced = res_g1_reduced[, c("log2FoldChange", "pvalue", "padj")]      
  
      n_DE = sum(res_g1_reduced$padj < 0.05 & abs(res_g1_reduced$log2FoldChange) > log2(1.5), 
                 na.rm = TRUE)
      print(paste0(d_group, " under KO gene ", geno, ":"))
      print(sprintf("number of DE genes: %d", n_DE))

      g1 = ggplot(res_g1_reduced %>% subset(!is.na(padj)), aes(x=pvalue)) + 
        geom_histogram(color="darkblue", fill="lightblue", 
                       breaks = seq(0,1,by=0.01)) + 
        ggtitle(paste0(d_group, "\n", geno))
      print(g1)

      tag1 = sprintf("%s_PTC_", geno)
      colnames(res_g1) = paste0(tag1, colnames(res_g1))
      colnames(res_g1_reduced) = paste0(tag1, colnames(res_g1_reduced))
      
      
      res_df = data.frame(gene_ID=rownames(res_g1), 
                          res_g1)
      dim(res_df)
      res_df[1:2,]
      
      
      res_reduced_gene_anno = gene_anno[match(rownames(res_g1_reduced), 
                                              gene_anno$ensembl_gene_id), ]
      stopifnot(all(rownames(res_g1_reduced)==res_reduced_gene_anno$ensembl_gene_id))     
      
      # set gene hgnc_symbol that equal "" to NA
      hgnc_symbol_vec = res_reduced_gene_anno$hgnc_symbol
      hgnc_symbol_vec[which(hgnc_symbol_vec=="")] = NA
      
      res_reduced_df = data.frame(gene_ID=rownames(res_g1_reduced), 
                                  hgnc_symbol=hgnc_symbol_vec,
                                  chr=res_reduced_gene_anno$chr, 
                                  strand=res_reduced_gene_anno$strand, 
                                  start=res_reduced_gene_anno$start, 
                                  end=res_reduced_gene_anno$end, 
                                  res_g1_reduced)
      
      print("hgnc_symbol empty string and NA tables:")
      print(table(res_reduced_df$hgnc_symbol=="", useNA="ifany"))
      print(table(is.na(res_reduced_df$hgnc_symbol)))
        
      dim(res_reduced_df)
      res_reduced_df[1:2,]      
      
      if (model1=="PrS"){
        setting_name = unique(paste0(sample_info_dg$model_day, "_", 
                                     sample_info_dg$condition))
      }else{
        setting_name = unique(sample_info_dg$model_day)
      }
      
      fwrite(res_df, 
             sprintf("%s/%s_%s_%s_DEseq2_raw.tsv", 
                     raw_output_dir, release_name, setting_name, geno), 
             sep="\t")
      fwrite(res_reduced_df, 
             sprintf("%s/%s_%s_%s_DEseq2.tsv", 
                     processed_output_dir, release_name, setting_name, geno), 
             sep="\t")      

    }
    
  }
  
}

colnames(df_size) = c("DE_group", "knockout_gene_strategy", "n_ko", "n_WT")
df_size = as.data.frame(df_size)

multi_rows = which(table(df_size$knockout_gene_strategy)>1)
multi_rows

df_size[which(df_size$knockout_gene_strategy%in%names(multi_rows)),]

fwrite(df_size, 
       sprintf("%s/%s_DE_n_samples.tsv", output_dir, release_name), 
       sep="\t")
```


## DE analysis with respect to condition (hypoxia vs. normoxia)

This is only relevant for model system PrS. 

Within model system PrS, for WT samples, use all WT samples (3 hypoxia for ASCL2_ELF5_HIF1A_TP63_hyp, 8 normoxia for ASCL2_ELF5_NR2F2_nor, DLX3_MSX2_VGLL1_nor, HIF1A_TP63_nor) for this analysis. 

Samples invoved: 11 WT, 6 ASCL2, 6 ELF5, 6 HIF1A, 6 TP63. For KO samples, half of them are under normoxia and the other half are under hypoxia. 

For the 11 WT samples, run regular DESeq2 on them.

```{r}
meta |> filter(model_system == 'PrS', ko_gene == 'WT') |> select(starts_with('clonal'), condition)
```

For HIF1A and TP63, their 6 samples are from three single cell clones, with each clone having two groups of cells developed under different conditions. 
Run paired DESeq2 on them. 

```{r}
meta |> filter(ko_gene == 'HIF1A') |> select(starts_with('clonal'), condition)
meta |> filter(ko_gene == 'TP63') |> select(starts_with('clonal'), condition)
```

For ASCL2 and ELF5, run regular DESeq2 on them.

```{r}
meta |> filter(ko_gene == 'ASCL2') |> select(starts_with('clonal'), condition)
meta |> filter(ko_gene == 'ELF5') |> select(starts_with('clonal'), condition)
```

```{r fig.width = 4.8, fig.height=3.2, message=FALSE, warning=FALSE}

df_size = NULL

meta$ko_group = paste(meta$ko_gene, meta$ko_strategy, sep="_")
relevant_ko_groups = unique(meta$ko_group[which(meta$condition=="hyp")])
relevant_ko_groups = sort(relevant_ko_groups)

for (cur_ko_group in relevant_ko_groups){
  
  sample2kp = which((meta$model_system == "PrS")&
                     (meta$ko_group == cur_ko_group))
  cts_dat_cond = cts_dat[,sample2kp]
  meta_cond    = meta[sample2kp,]
  
  dim(cts_dat_cond)
  dim(meta_cond)
  
  unique_model_day = unique(paste0(meta_cond$model_system, "_day_", 
                                   meta_cond$timepoint))
  print("")
  print("==========================================")
  print(paste0("PrS day 6, ", cur_ko_group, " samples"))
  print("hyp vs nor DE results:")
  print("==========================================")
  print("")
  
  stopifnot(all(meta_cond$file_id==colnames(cts_dat_cond)))
  table(meta_cond$file_id==colnames(cts_dat_cond), useNA="ifany")
  stopifnot(all(meta_cond$ko_group==cur_ko_group))
  
  print(table(meta_cond$condition, meta_cond$run_id_short))
  
  cols2kp = c("model_system", "timepoint", "clonal.label", "condition")
  sample_info_cond = meta_cond[,cols2kp]
  colnames(sample_info_cond)[which(colnames(sample_info_cond)=="clonal.label")] = "clonal_label"
  
  dim(sample_info_cond)
  sample_info_cond[1:2,]
  
  # do two steps of filtering
  # first, filter based on q75 of each gene
  # second, filter based on whether each gene is expressed in at least 4 cells
  
  cond_q75 = apply(cts_dat_cond, 1, quantile, probs=0.75)
  cts_dat_cond = cts_dat_cond[cond_q75 > 0,]
  
  print(sprintf("After filtering by gene expression q75, the number of genes reduces from %d to %d", 
                length(cond_q75), nrow(cts_dat_cond)))
  
  n_pos = rowSums(cts_dat_cond>0)
  cts_dat_cond = cts_dat_cond[n_pos >= 4,]
  
  if (length(n_pos)==nrow(cts_dat_cond)){
    print("After filtering by nonzero expression in at least 4 samples, the number of genes does not change.")
  }else{
    print(sprintf("After filtering by nonzero expression in at least 4 samples, the number of genes reduces from %d to %d", 
                  length(n_pos), nrow(cts_dat_cond)))    
  }
  
  # update the q75 based on only genes that will be used in the DE analysis
  sample_info_cond$q75 = apply(cts_dat_cond, 2, quantile, probs = 0.75)
  sample_info_cond$log_q75 = log(sample_info_cond$q75)
  
  if (cur_ko_group %in% c("WT_WT", "ASCL2_PTC", "ELF5_PTC")){
    dds = DESeqDataSetFromMatrix(cts_dat_cond, colData=sample_info_cond, 
                                  ~ condition + log_q75)
  }else{
    dds = DESeqDataSetFromMatrix(cts_dat_cond, colData=sample_info_cond, 
                                  ~ clonal_label + log_q75 + condition)     
  }
  
  start.time <- Sys.time()
  dds = DESeq(dds)
  end.time <- Sys.time()
  
  time.taken <- end.time - start.time
  print(time.taken)
  
  ## association with read-depth
  res_rd = results(dds, name = "log_q75")
  res_rd = as.data.frame(res_rd)
  dim(res_rd)
  res_rd[1:2,]
  
  table(is.na(res_rd$padj))
  
  g0 = ggplot(res_rd %>% subset(!is.na(padj)), aes(x=pvalue)) + 
    geom_histogram(color="darkblue", fill="lightblue", 
                   breaks = seq(0,1,by=0.01)) + 
    ggtitle(paste0("PrS day 6 hyp vs nor ", cur_ko_group, "\nread depth"))
  print(g0)
  
  ## Rerun the analysis without read-depth if it is not significant for 
  ## most genes, and the number of samples is small
  ## i.e., proportion of non-null in the distribution is smaller than 0.01
  ## (this needs to be restricted to the genes with not NA adjusted pvalue)
  ## or if smaller than 0.1 and the number of samples is not greater than 6
  
  pi_1_rd = max(0, mean(res_rd$pvalue[!is.na(res_rd$padj)] < 0.05) - 0.05)
  pi_1_rd = max(pi_1_rd, 0, 1 - 2*mean(res_rd$pvalue[!is.na(res_rd$padj)] > 0.5))
  
  print(sprintf("prop of non-null for rd: %.2e", pi_1_rd))
  
  if(pi_1_rd < 0.01 || (ncol(cts_dat_cond) <= 6 && pi_1_rd < 0.1 )){
    print("rerun DESeq2 without read depth")
    if (cur_ko_group=="WT_WT"){
      dds = DESeqDataSetFromMatrix(cts_dat_cond, colData=sample_info_cond, 
                                    ~ condition)
    }else{
      dds = DESeqDataSetFromMatrix(cts_dat_cond, colData=sample_info_cond, 
                                    ~ clonal_label + condition)  
    }
    dds = DESeq(dds)
  }
  
  ## DE analysis between two conditions
  res_g1 = results(dds, contrast = c("condition", "hyp", "nor"))
  res_g1 = signif(data.frame(res_g1), 3)
  
  # add a column to record the number of zeros per test
  n_zero = rowSums(cts_dat_cond==0)
  res_g1$n_zeros = n_zero    
  
  # record the number of samples involved in the comparison
    
  df_size = rbind(df_size, 
                  c(unique_model_day, 
                    cur_ko_group, 
                    sum(sample_info_cond$condition=="hyp"), 
                    sum(sample_info_cond$condition=="nor")))   
    
  # prepare a processed version of output
  res_g1_reduced = res_g1[, c("log2FoldChange", "pvalue", "padj", "n_zeros")]
  res_g1_reduced$padj[which(res_g1_reduced$n_zeros>=4)] = NA
  
  res_g1_reduced = res_g1_reduced[, c("log2FoldChange", "pvalue", "padj")]      
  
  n_DE = sum(res_g1_reduced$padj < 0.05 & abs(res_g1_reduced$log2FoldChange) > log2(1.5), 
             na.rm = TRUE)
  print(paste0("PrS between nor and hyp conditions, ", cur_ko_group, ":"))
  print(sprintf("number of DE genes: %d", n_DE))
  
  g1 = ggplot(res_g1_reduced %>% subset(!is.na(padj)), aes(x=pvalue)) + 
    geom_histogram(color="darkblue", fill="lightblue", 
                   breaks = seq(0,1,by=0.01)) + 
    ggtitle(paste0("PrS day 6 hyp vs nor\n", cur_ko_group))
  print(g1)
  
  tag1 = "hyp_vs_nor_"
  colnames(res_g1) = paste0(tag1, colnames(res_g1))
  colnames(res_g1_reduced) = paste0(tag1, colnames(res_g1_reduced))
  
  
  res_df = data.frame(gene_ID=rownames(res_g1), 
                      res_g1)
  dim(res_df)
  res_df[1:2,]
    
    
  res_reduced_gene_anno = gene_anno[match(rownames(res_g1_reduced), 
                                          gene_anno$ensembl_gene_id), ]
  stopifnot(all(rownames(res_g1_reduced)==res_reduced_gene_anno$ensembl_gene_id))     
    
  # set gene hgnc_symbol that equal "" to NA
  hgnc_symbol_vec = res_reduced_gene_anno$hgnc_symbol
  hgnc_symbol_vec[which(hgnc_symbol_vec=="")] = NA
    
  res_reduced_df = data.frame(gene_ID=rownames(res_g1_reduced), 
                              hgnc_symbol=hgnc_symbol_vec,
                              chr=res_reduced_gene_anno$chr, 
                              strand=res_reduced_gene_anno$strand, 
                              start=res_reduced_gene_anno$start, 
                              end=res_reduced_gene_anno$end, 
                              res_g1_reduced)
  
  print("hgnc_symbol empty string and NA tables:")
  print(table(res_reduced_df$hgnc_symbol=="", useNA="ifany"))
  print(table(is.na(res_reduced_df$hgnc_symbol)))
      
  dim(res_reduced_df)
  res_reduced_df[1:2,]      
    
  fwrite(res_df, 
         sprintf("%s/%s_%s_%s_DEseq2_raw.tsv", 
                 raw_output_dir, release_name, 
                 paste0("PrS_day_6_", cur_ko_group), "hyp_vs_nor"), 
         sep="\t")
  fwrite(res_reduced_df, 
         sprintf("%s/%s_%s_%s_DEseq2.tsv", 
                 processed_output_dir, release_name, 
                 paste0("PrS_day_6_", cur_ko_group), "hyp_vs_nor"), 
         sep="\t")  

}

colnames(df_size) = c("model_system_day", "knockout_gene_strategy", "n_hyp", "n_nor")
fwrite(df_size, 
       sprintf("%s/%s_DE_hyp_vs_nor_n_samples.tsv", output_dir, release_name), 
       sep="\t")

```



```{r}
gc()
sessionInfo()
```
